packages/flutter/lib/src/rendering/viewport.dart

```diff
@@ -319,7 +380,60 @@ abstract class RenderViewportBase<ParentDataClass extends ContainerParentDataMix
     }
 
     // we made it without a correction, whee!
-    return 0.0;
+    return  0.0;
+  }
+
+  @override
+  Rect describeApproximatePaintClip(RenderSliver child) {
+    final Rect viewportClip = Offset.zero & size;
+    if (child.constraints.overlap == 0) {
+      return viewportClip;
+    }
+
+    // Adjust the clip rect for this sliver by the overlap from the previous sliver.
+    double left = viewportClip.left;
+    double right = viewportClip.right;
+    double top = viewportClip.top;
+    double bottom = viewportClip.bottom;
+    final double startOfOverlap = child.constraints.viewportMainAxisExtent - child.constraints.remainingPaintExtent;
+    final double overlapCorrection = startOfOverlap + child.constraints.overlap;
+    switch (applyGrowthDirectionToAxisDirection(axisDirection, child.constraints.growthDirection)) {
+      case AxisDirection.down:
+        top += overlapCorrection;
+        break;
+      case AxisDirection.up:
+        bottom -= overlapCorrection;
+        break;
+      case AxisDirection.right:
+        left += overlapCorrection;
+        break;
+      case AxisDirection.left:
+        right -= overlapCorrection;
+        break;
+    }
+    return new Rect.fromLTRB(left, top, right, bottom);
+  }
+
+  @override
+  Rect describeSemanticsClip(RenderSliver child) {
+    switch(axis) {
+      case Axis.vertical:
+        return new Rect.fromLTRB(
+          semanticBounds.left,
+          semanticBounds.top - cacheExtent,
```

Comment:
(same comment in the other instance of this method):

Why a fixed extend out in the scrolling axis, instead of near-infinity?

Reply:
To keep the semantics tree compact.

In the context of the SingleChildViewport: Imagine a thousands of pixels long column inside the viewport with many child SemanticsNodes (one per cell). Only the semantics nodes that fall inside the visible range plus a handful of objects before and after the viewport are relevant. Everything else is overhead. This ensures we cut those other child semantic nodes out.

Same is true for the other Sliver-based viewport: Imagine a loooong sliver with many child semantics nodes. Again, only some of these are relevant.